### PR TITLE
feat: svelte query client (#683)Co-authored-by: Sergiy Petrunin <spetrunin@users.noreply.github.com> Co-authored-by: Eelco Wiersma <contact@pagebakers.nl> Co-authored-by: Pagebakers <eelco@appulse.nl>

### PR DESCRIPTION
**Pull request by DaniAkash**
This PR is a draft of my initial implementation of [Wundergraph Client for sveltekit](https://github.com/DaniAkash/wundergraph-svelte-query), created to collect feedback from the maintainers of this project. There's still few pending items I need to address before this PR can be ready:
- [x] figure out how to create subscription utilities in svelte
- [x] setup tests for the svelte query client utilities
  - [x] subscriptions broken - need to work on a fix
- [ ] move [Wundergraph Client for sveltekit](https://github.com/DaniAkash/wundergraph-svelte-query) into examples

#### Checklist

- [ ] run `make test`
- [x] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Pagebakers** [comment]
Great work! is there anything you need help with?

**Comment by DaniAkash** [comment]
> Great work! is there anything you need help with?

The subscription test cases are failing. Might need some help with it.

**Comment by DaniAkash** [comment]
Since I'm new to svelte, I might have made some mistakes in the readable store implementation. Will try to figure out the issue tonight

**Comment by Pagebakers** [comment]
Thanks, I will try it out once I have the time.

**Comment by DaniAkash** [comment]
> Thanks, I will try it out once I have the time.

It is fixed now. All tests are passing 🎉

File: packages/svelte-query/README.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    <slot />
+  </QueryClientProvider>
+</div>
+```
+
+Now you can use svelte-query to call your wundergraph operations!
+
+```svelte
+<script lang="ts">
+	import { createQuery } from '../lib/svelteQueryUtils';
```
Comment (Pagebakers) [inline packages/svelte-query/README.md:77]: I'd call it `lib/wundergraph` to make it consistent with the other query libs.

File: packages/svelte-query/README.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+```ts
+import { createQueryUtils } from '@wundergraph/svelte-query';
+import { createClient } from '../generated/client';
+import type { Operations } from '../generated/client';
+
+const client = createClient(); // Typesafe WunderGraph client
+
+// These utility functions needs to be imported into your app
+export const { createQuery, createFileUpload, createMutation, createSubscription, getAuth, getUser, queryKey } =
+  createQueryUtils<Operations>(client);
```
Comment (Pagebakers) [inline packages/svelte-query/README.md:45]: Calling this Utils is a Svelte convention? I'm not sure.

File: packages/svelte-query/package.json [hunk 3]
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "@wundergraph/svelte-query",
+  "version": "0.0.1",
+  "license": "Apache-2.0",
+  "description": "WunderGraph Svelte Query Integration",
+  "main": "./dist/index.js",
```
Comment (Pagebakers) [inline packages/svelte-query/package.json:6]: We should release this as a dual bundle with exports (commonjs/esm), you can copy the tsup script and exports + main/module/types fields from the react-query package.

File: packages/svelte-query/src/createQueryUtils.ts [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+			},
+		};
+	};
+
+	/**
+	 * Return the logged in user.
+	 *
+	 * @usage
+	 * ```ts
+	 * const { user, error, isLoading } = getUser()
```
Comment (Pagebakers) [inline packages/svelte-query/src/createQueryUtils.ts:145]: `const { data`

File: packages/svelte-query/src/createQueryUtils.ts [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+			liveQuery,
+			enabled: options.enabled !== false && liveQuery,
+			onSuccess: options.onSuccess,
+			onError: options.onError,
+		});
+
+		if (liveQuery) {
+			return {
+				...result,
+				subscriptionState,
```
Comment (Pagebakers) [inline packages/svelte-query/src/createQueryUtils.ts:95]: deconstruct this into the return value. 

```
return {
  ...result,
   ...subscriptionState
}

const { isSubscribed, isLoading } = useQuery()
```

I'm not sure how reactivity works with svelte but for Solid we had to use `mergeProps` to keep the reactivity.

File: packages/svelte-query/src/createQueryUtils.ts [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+			input,
+			subscribeOnce,
+			enabled,
+			onSuccess,
+			onError,
+		});
+
+		return {
+			...subscription,
+			subscriptionState,
```
Comment (Pagebakers) [inline packages/svelte-query/src/createQueryUtils.ts:297]: Same as livequery deconstruct this into the return value.

File: packages/svelte-query/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
+{
+  "extends": "../../tsconfig.base.json",
+  "compilerOptions": {
+    "target": "ES2019",
+    "module": "commonjs",
+    "moduleResolution": "node",
+    "declarationDir": "./dist",
+    "jsx": "react",
```
Comment (Pagebakers) [inline packages/svelte-query/tsconfig.json:8]: This doesn't seem correct I think?

File: packages/svelte-query/tsconfig.build.json [hunk 8]
```diff
@@ -0,0 +1,4 @@
+{
+  "extends": "./tsconfig.json",
+  "include": ["src/**/*"]
```
Comment (Pagebakers) [inline packages/svelte-query/tsconfig.build.json:3]: For the build config I suggest we use the same setup as tanstack svelte-query

**Review by Pagebakers — CHANGES_REQUESTED** [review]
Great! Left some comments for improvements.

Will you also add a svelte/sveltekit example?

File: packages/svelte-query/README.md (hunk 2 above)
Comment (DaniAkash) [reply packages/svelte-query/README.md:45]: no, couldn't find any common name like hooks in react

File: packages/svelte-query/package.json (hunk 3 above)
Comment (DaniAkash) [reply packages/svelte-query/package.json:6]: Pretty much all svelte setups (vite / sveltekit) support ESM so conventionally svelte packages don't export UMD or CJS builds 

Refer: 
https://github.com/TanStack/query/discussions/5006
https://github.com/TanStack/query/pull/4768#issuecomment-1374622732

File: packages/svelte-query/src/createQueryUtils.ts (hunk 5 above)
Comment (DaniAkash) [reply packages/svelte-query/src/createQueryUtils.ts:95]: Will figure out some way to preserve reactivity

File: packages/svelte-query/tsconfig.json (hunk 7 above)
Comment (DaniAkash) [reply packages/svelte-query/tsconfig.json:8]: I think JSX is not needed, I'll change the tsconfig to svelte query's config

File: packages/svelte-query/README.md (hunk 1 above)
Comment (DaniAkash) [reply packages/svelte-query/README.md:77]: > Great overall! Left some comments for improvements.
> 
> Will you also add a svelte/sveltekit example?

I'd like to add, just waiting for the package to be published to npm

Comment (Pagebakers) [reply packages/svelte-query/README.md:77]: Good point

File: packages/svelte-query/README.md (hunk 2 above)
Comment (Pagebakers) [reply packages/svelte-query/README.md:45]: How about `createSvelteClient`

File: packages/svelte-query/package.json (hunk 3 above)
Comment (Pagebakers) [reply packages/svelte-query/package.json:6]: Cool, in that case the package type needs to be module. And I'd still use ts-up for the build, but than only output esm, as well as using exports in package.json

**Review by Pagebakers — CHANGES_REQUESTED** [review]

File: packages/svelte-query/README.md (hunk 2 above)
Comment (DaniAkash) [reply packages/svelte-query/README.md:45]: this sounds much better. will refactor 👍

File: packages/svelte-query/tsconfig.json (hunk 7 above)
Comment (DaniAkash) [reply packages/svelte-query/tsconfig.json:8]: have updated the config based on svelte query

File: packages/svelte-query/package.json (hunk 3 above)
Comment (DaniAkash) [reply packages/svelte-query/package.json]: Sveltekit docs recommended using [svelte-package](https://kit.svelte.dev/docs/packaging) for building packages, Tanstack svelte-query also used the same build tool, so changed the build config as per the svelte-package docs

File: packages/svelte-query/src/createQueryUtils.ts (hunk 5 above)
Comment (DaniAkash) [reply packages/svelte-query/src/createQueryUtils.ts]: @Pagebakers have use derived to merge the queryResult & subscription state, will look for ways to improve the typings on the derived state. But everything looks good for now

**Comment by DaniAkash** [comment]
> Great overall! Left some comments for improvements.
> 
> Will you also add a svelte/sveltekit example?

@Pagebakers would like your review on the derived state implementation. If it looks good, I'll work on improving the types

File: packages/svelte-query/src/lib/createSvelteClient.ts [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+			operationName,
+			input,
+			liveQuery,
+			enabled: options.enabled !== false && liveQuery,
+			onSuccess: options.onSuccess,
+			onError: options.onError,
+		});
+
+		if (liveQuery) {
+			const queryWithSubscriptionState: Readable<
```
Comment (Pagebakers) [inline packages/svelte-query/src/lib/createSvelteClient.ts:95]: Looks good @DaniAkash!

I'm thinking we could re-use this derived code and turn it into to a helper.

For example:
```
return withSubscriptionState(result, subscriptionState)
```

Comment (DaniAkash) [reply packages/svelte-query/src/lib/createSvelteClient.ts]: Have created the helper along with the correct types

**Comment by Pagebakers** [comment]
Code looks good, but I'm unable to run the tests locally. (with CI it's expected not to work)

```
 FAIL  tests/queryUtils.test.ts (8.807 s)
  Svelte Query - createSvelteClient
    ✓ should be able to init utility functions (1 ms)
    ✕ should return data (1035 ms)
    ✓ should be disabled (182 ms)
  Svelte Query - createMutation
    ✓ should trigger mutation with auth (17 ms)
    ✓ should trigger mutation (8 ms)
    ✕ should invalidate query (1010 ms)
  Svelte Query - createSubscription
    ✕ should subscribe (5006 ms)
  Svelte Query - getUser
    ✕ should return user (1021 ms)
    ```

**Review by Pagebakers — APPROVED** [review]

**Pull request merged**